core/go/internal/sequencer/coordinator/coordinator.go

```diff
@@ -0,0 +1,600 @@
+		err = c.transactionsByID[txn.ID].HandleEvent(ctx, receivedEvent)
+		if err != nil {
+			log.L(ctx).Errorf("error handling ReceivedEvent for transaction %s: %v", txn.ID.String(), err)
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *coordinator) queueForDispatch(ctx context.Context, txn *transaction.Transaction) {
+	c.dispatchQueue <- txn
+}
+
+func (c *coordinator) propagateEventToTransaction(ctx context.Context, event transaction.Event) error {
+	if txn := c.transactionsByID[event.GetTransactionID()]; txn != nil {
+		return txn.HandleEvent(ctx, event)
+	} else {
+		log.L(ctx).Debugf("ignoring event because transaction not known to this coordinator %s", event.GetTransactionID().String())
+	}
+	return nil
+}
+
+func (c *coordinator) propagateEventToAllTransactions(ctx context.Context, event common.Event) error {
+	for _, txn := range c.transactionsByID {
+		err := txn.HandleEvent(ctx, event)
+		if err != nil {
+			log.L(ctx).Errorf("error handling event %v for transaction %s: %v", event.Type(), txn.ID.String(), err)
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *coordinator) getTransactionsInStates(ctx context.Context, states []transaction.State) []*transaction.Transaction {
+	//TODO this could be made more efficient by maintaining a separate index of transactions for each state but that is error prone so
+	// deferring until we have a comprehensive test suite to catch errors
+	log.L(ctx).Debugf("getting transactions in states: %+v", states)
+	matchingStates := make(map[transaction.State]bool)
+	for _, state := range states {
+		matchingStates[state] = true
+	}
+
+	log.L(ctx).Debugf("looking across %d transactions for those in states: %+v", len(c.transactionsByID), states)
+	matchingTxns := make([]*transaction.Transaction, 0, len(c.transactionsByID))
+	for _, txn := range c.transactionsByID {
+		if matchingStates[txn.GetState()] {
+			log.L(ctx).Debugf("found transaction %s in state %s", txn.ID.String(), txn.GetState())
+			matchingTxns = append(matchingTxns, txn)
+		}
+	}
+	return matchingTxns
+}
+
+func (c *coordinator) getTransactionsNotInStates(ctx context.Context, states []transaction.State) []*transaction.Transaction {
+	//TODO this could be made more efficient by maintaining a separate index of transactions for each state but that is error prone so
+	// deferring until we have a comprehensive test suite to catch errors
+	nonMatchingStates := make(map[transaction.State]bool)
+	for _, state := range states {
+		nonMatchingStates[state] = true
+	}
+	matchingTxns := make([]*transaction.Transaction, 0, len(c.transactionsByID))
+	for _, txn := range c.transactionsByID {
+		if !nonMatchingStates[txn.GetState()] {
+			matchingTxns = append(matchingTxns, txn)
+		}
+	}
+	return matchingTxns
+}
+
+// MRW TODO - is there a reason we need to find by nonce and not by TX ID?
+func (c *coordinator) findTransactionBySignerNonce(ctx context.Context, signer *pldtypes.EthAddress, nonce uint64) *transaction.Transaction {
```

Comment:
What's the scenario for this function?

I can imagine scenarios like "Maybe this public transaction was submitted as part of this private transaction", but it's not clear to me in isolation here.